iree/hal/local/elf/elf_module.c

```diff
@@ -0,0 +1,630 @@
+// of the ELF but is yet unadjusted for host requirements. The range will have
+// zero length if there are no segments to load (which would be weird).
+static iree_byte_range_t iree_elf_module_calculate_vaddr_range(
+    iree_elf_module_load_state_t* load_state) {
+  // Min/max virtual addresses of any allocated segment.
+  iree_elf_addr_t vaddr_min = IREE_ELF_ADDR_MAX;
+  iree_elf_addr_t vaddr_max = IREE_ELF_ADDR_MIN;
+  for (iree_elf_half_t i = 0; i < load_state->ehdr->e_phnum; ++i) {
+    const iree_elf_phdr_t* phdr = &load_state->phdr_table[i];
+    if (phdr->p_type != IREE_ELF_PT_LOAD) continue;
+    iree_elf_addr_t p_vaddr_min =
+        iree_page_align_start(phdr->p_vaddr, phdr->p_align);
+    iree_elf_addr_t p_vaddr_max =
+        iree_page_align_end(phdr->p_vaddr + phdr->p_memsz, phdr->p_align);
+    vaddr_min = iree_min(vaddr_min, p_vaddr_min);
+    vaddr_max = iree_max(vaddr_max, p_vaddr_max);
+  }
+  if (vaddr_min == IREE_ELF_ADDR_MAX) {
+    // Did not find any segments to load.
+    vaddr_min = IREE_ELF_ADDR_MIN;
+    vaddr_max = IREE_ELF_ADDR_MIN;
+  }
+  iree_byte_range_t byte_range = {
+      .offset = (iree_host_size_t)vaddr_min,
+      .length = (iree_host_size_t)(vaddr_max - vaddr_min),
+  };
+  return byte_range;
+}
+
+// Allocates space for and loads all DT_LOAD segments into the host virtual
+// address space.
+static iree_status_t iree_elf_module_load_segments(
+    iree_const_byte_span_t raw_data, iree_elf_module_load_state_t* load_state,
+    iree_elf_module_t* module) {
+  // Calculate the total internally-aligned vaddr range.
+  iree_byte_range_t vaddr_range =
+      iree_elf_module_calculate_vaddr_range(load_state);
+
+  // Reserve virtual address space in the host memory space. This memory is
+  // uncommitted by default as the ELF may only sparsely use the address space.
+  module->vaddr_size = iree_page_align_end(
+      vaddr_range.length, load_state->memory_info.normal_page_size);
+  IREE_RETURN_IF_ERROR(iree_memory_view_reserve(
+      IREE_MEMORY_VIEW_FLAG_MAY_EXECUTE, module->vaddr_size,
+      (void**)&module->vaddr_base));
+  module->vaddr_bias = module->vaddr_base - vaddr_range.offset;
+
+  // Commit and load all of the segments.
+  for (iree_elf_half_t i = 0; i < load_state->ehdr->e_phnum; ++i) {
+    const iree_elf_phdr_t* phdr = &load_state->phdr_table[i];
+    if (phdr->p_type != IREE_ELF_PT_LOAD) continue;
+
+    // Commit the range of pages used by this segment, initially with write
+    // access so that we can modify the pages.
+    iree_byte_range_t byte_range = {
+        .offset = phdr->p_vaddr,
+        .length = phdr->p_memsz,
+    };
+    IREE_RETURN_IF_ERROR(iree_memory_view_commit_ranges(
+        module->vaddr_bias, 1, &byte_range,
+        IREE_MEMORY_ACCESS_READ | IREE_MEMORY_ACCESS_WRITE));
+
+    // Copy data present in the file.
+    if (phdr->p_filesz > 0) {
+      memcpy(module->vaddr_bias + phdr->p_vaddr, raw_data.data + phdr->p_offset,
```

Comment:
Comment that as a future optimization, we can overlay the memory from fixed mmaps (on archs where this is convenient and an actual optimization)?